packages/create-subscription/src/createSubscription.js

```diff
@@ -0,0 +1,156 @@
+/**
+ * Copyright (c) 2014-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import React from 'react';
+import warning from 'fbjs/lib/invariant';
+
+export function createSubscription<
+  Property,
+  CreatedSubscription,
+  Value,
+>(config: {|
+  // Synchronously gets the value for the subscribed property.
+  // Return undefined if the subscribable value is undefined,
+  // Or does not support synchronous reading (e.g. native Promise).
+  +getValue: (source: Property) => Value,
+
+  // Setup a subscription for the subscribable value in props.
+  // Due to the variety of change event types, subscribers should provide their own handlers.
+  // Those handlers should not attempt to update state though;
+  // They should call the callback() instead when a subscription changes.
+  // You may optionally return a subscription value to later unsubscribe (e.g. event handler).
+  +subscribe: (
+    source: Property,
+    callback: (value: Value) => void,
+  ) => CreatedSubscription,
+
+  // Unsubsribe from the subscribable value in props.
+  // The subscription value returned from subscribe() is passed as the second parameter.
+  +unsubscribe: (source: Property, subscription: CreatedSubscription) => void,
+|}): React$ComponentType<{
+  children: (value: Value) => React$Element<any>,
```

Comment:
`Value` could be `undefined` here if `props.source` is. Should probably be `Value | void`? Would be nice if the flow type could reflect that `Value` is always set for the case that the subscription can provide that and the `source` is non-nullable.

Reply:
Cool. I'll update to `Value | void`